localstack/services/logs/provider.py

```diff
@@ -2,101 +2,156 @@
-from localstack.utils.patch import Patches
+from localstack.utils.common import is_number
+from localstack.utils.patch import patch
 
 LOG = logging.getLogger(__name__)
 
 
-class ProxyListenerCloudWatchLogs(ProxyListener):
-    def forward_request(self, method, path, data, headers):
-        action = headers.get("X-Amz-Target") or ""
-        action = action.split(".")[-1]
-
-        if action == "PutLogEvents":
-            publish_log_metrics_for_events(data)
-
-        if method == "POST" and path == "/":
-            if "nextToken" in to_str(data or ""):
-                data = self._fix_next_token_request(data)
-                headers["Content-Length"] = str(len(data))
-                return Request(data=data, headers=headers, method=method)
-
-        return True
+class LogsProvider(LogsApi, ABC):
+    def __init__(self):
+        super().__init__()
+        self.cw_client = aws_stack.connect_to_service("cloudwatch")
 
-    def return_response(self, method, path, data, headers, response):
-        # Fix Incorrect response content-type header from cloudwatch logs #1343
+    def put_log_events(
+        self,
+        context: RequestContext,
+        log_group_name: LogGroupName,
+        log_stream_name: LogStreamName,
+        log_events: InputLogEvents,
+        sequence_token: SequenceToken = None,
+    ) -> PutLogEventsResponse:
+        logs_backend = logs_backends[aws_stack.get_region()]
+        metric_filters = logs_backend.filters.metric_filters
+        for metric_filter in metric_filters:
+            pattern = metric_filter.get("filterPattern", "")
+            transformations = metric_filter.get("metricTransformations", [])
+            matches = get_pattern_matcher(pattern)
+            for log_event in log_events:
+                if matches(pattern, log_event):
+                    for tf in transformations:
+                        value = tf.get("metricValue") or "1"
+                        if "$size" in value:
+                            LOG.info(
+                                "Expression not yet supported for log filter metricValue", value
+                            )
+                        value = float(value) if is_number(value) else 1
+                        data = [{"MetricName": tf["metricName"], "Value": value}]
+                        try:
+                            self.cw_client.put_metric_data(
+                                Namespace=tf["metricNamespace"], MetricData=data
+                            )
+                        except Exception as e:
+                            LOG.info(
+                                "Unable to put metric data for matching CloudWatch log events", e
+                            )
+        return call_moto(context)
+
+
+class LogsAwsApiListener(AwsApiListener):
+    def __init__(self):
+        self.provider = LogsProvider()
+        super().__init__("logs", MotoFallbackDispatcher(self.provider))
+
+    def return_response(
```

Comment:
Let's better use the new interface of `AwsApiListener` here, as we're gradually moving away from the older/legacy interface of `ProxyListener` (i.e., `forward_request`/`return_response`).

Something like this should work:
```
    def request(self, request: Request) -> Response:
        response = super().request(request)
        response.headers["content-type"] = APPLICATION_AMZ_JSON_1_1
        return response
```